src/image.cpp

```diff
@@ -389,20 +402,27 @@ namespace Exiv2 {
                 // if ( offset > io.size() ) offset = 0; // Denial of service?
 
                 // #55 and #56 memory allocation crash test/data/POC8
-                long long allocate = static_cast<long long>(size) * count + pad + 20;
-                if (allocate > static_cast<long long>(io.size())) {
+                const uint64_t allocate64 = static_cast<uint64_t>(size) * count + pad + 20;
+                if ( allocate64 > io.size() ) {
                     throw Error(kerInvalidMalloc);
                 }
-                DataBuf buf(static_cast<long>(allocate));  // allocate a buffer
+                // Overflow check
+                enforce(allocate64 <= static_cast<uint64_t>(std::numeric_limits<uint32_t>::max()), kerCorruptedMetadata);
+                enforce(allocate64 <= static_cast<uint64_t>(std::numeric_limits<long>::max()), kerCorruptedMetadata);
```

Comment:
@kevinbackhouse there is a LGTM warning reported for this line (in #1778, as well, but I missed it there... :facepalm: )      
Maybe it's better not to use `long` given that this will be a different size on Win vs Linux? (https://en.cppreference.com/w/cpp/language/types)

If `allocate64` is calculated as above, and then checked that it's smaller than `io.size()`, isn't that enough?  
`io.size()` is `size_t` so isn't there a scenario where `allocate64` is greater than `std::numeric_limits<uint32_t>::max()` and that would be legal? 

It seems that we really need more tests that run this code... I'll try and have a look if we can improve coverage here

Reply:
@hassec: I was aware of the LGTM warning but decided to ignore it because I think we need this second `enforce` on Windows, where `long` is only 32-bits. The use of `long` is a problem that is quite deeply ingrained in the whole codebase. For example, [`basicio`](https://github.com/Exiv2/exiv2/blob/0208b508a0fe4079e146d43c44299a1ebcd480ee/include/exiv2/basicio.hpp#L144) and [`DataBuf`](https://github.com/Exiv2/exiv2/blob/0208b508a0fe4079e146d43c44299a1ebcd480ee/include/exiv2/types.hpp#L180) both use `long` for sizes, when something like `size_t` would probably have been a better choice. I think we need to try to systematically replace it, but it will be a fairly big project.

I added this second `enforce` to ensure that there won't be an integer overflow in the call to the constructor of `DataBuf` on line 413 or the call to `readOrThrow` on line 424. LGTM is warning about a pointless comparison because it is doing the analysis on Linux, where `long` is 64-bits.

Reply:
@kevinbackhouse I think I still don't quite understand all of this.   

On windows, `long` is 32-bits, which is just `uint32_t` so won't the two calls to `enforce` be identical on windows?
So we probably only need the first. 

And you've convinced me that given the use of `long`, we should probably enforce this. Even though from a logic standpoint, a directory could have a size > `std::numeric_limits<uint32_t>::max()`, you are right that some classes aren't ready for that on windows...      
It would probably make sense to turn the `long` -> `size_t` into an issue, so people are aware of it.
Would even be a good issue for newcomers, as one can do it step by step.

Reply:
Ah, found it. We already have an issue to track this problem: https://github.com/Exiv2/exiv2/issues/1538
I think `long` is equivalent to `int32_t` on Windows, so it has a lower max than `uint32_t`.

Reply:
thanks for pointing me to the issue! 

> I think long is equivalent to int32_t on Windows, so it has a lower max than uint32_t.  

right.... as explained on the cppreference that I linked... :facepalm:  sorry for the noise :sweat_smile: